cmd/object-handlers.go

```diff
@@ -1566,10 +1574,18 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 	}
 
 	// Check if bucket encryption is enabled
-	_, err = globalBucketSSEConfigSys.Get(bucket)
+	sseConfig, err := globalBucketSSEConfigSys.Get(bucket)
 	// This request header needs to be set prior to setting ObjectOptions
-	if (globalAutoEncryption || err == nil) && !crypto.SSEC.IsRequested(r.Header) {
-		r.Header.Set(xhttp.AmzServerSideEncryption, xhttp.AmzEncryptionKMS)
+	if _, ok := crypto.IsRequested(r.Header); !ok && (globalAutoEncryption || err == nil) {
+		switch {
+		case err == nil && sseConfig.Algorithm() == xhttp.AmzEncryptionAES:
+			r.Header.Set(xhttp.AmzServerSideEncryption, xhttp.AmzEncryptionAES)
+		case err == nil && sseConfig.Algorithm() == xhttp.AmzEncryptionKMS:
+			r.Header.Set(xhttp.AmzServerSideEncryption, xhttp.AmzEncryptionKMS)
+			r.Header.Set(xhttp.AmzServerSideEncryptionKmsID, sseConfig.KeyID())
+		default:
+			r.Header.Set(xhttp.AmzServerSideEncryption, xhttp.AmzEncryptionKMS)
+		}
```

Comment:
Looks like a good thing to be written in a single function..

Reply:
Yeah - actually let's make this a `SSEBucketConfig` method. One minute...